### PR TITLE
honour ALLOCA_MAX limit on all alloca() allocations, not just some

### PR DESCRIPTION
**Pull request by poettering**
Inspired by #19973

### CODE REVIEW

File: src/basic/alloc-util.h [hunk 1]
```diff
@@ -67,17 +72,15 @@ void* memdup_suffix0(const void *p, size_t l); /* We can't use _alloc_() here, s
         ({                                      \
                 void *_q_;                      \
                 size_t _l_ = l;                 \
-                assert(_l_ <= ALLOCA_MAX);      \
-                _q_ = alloca(_l_ ?: 1);         \
+                _q_ = alloca_safe(_l_ ?: 1);    \
```
Comment (keszybz) [inline src/basic/alloc-util.h]: `?: 1` not needed here.

File: src/basic/alloc-util.h [hunk 2]
```diff
@@ -155,8 +157,7 @@ void* greedy_realloc0(void **p, size_t need, size_t size);
                 void *_ptr_;                                            \
                 size_t _mask_ = (align) - 1;                            \
                 size_t _size_ = size;                                   \
-                assert(_size_ <= ALLOCA_MAX);                           \
-                _ptr_ = alloca((_size_ + _mask_) ?: 1);                 \
+                _ptr_ = alloca_safe((_size_ + _mask_));                 \
```
Comment (keszybz) [inline src/basic/alloc-util.h]: `()` not needed here.

File: src/basic/log.c [hunk 3]
```diff
@@ -1073,8 +1073,11 @@ int log_struct_iovec_internal(
 
         for (size_t i = 0; i < n_input_iovec; i++)
                 if (memory_startswith(input_iovec[i].iov_base, input_iovec[i].iov_len, "MESSAGE=")) {
-                        char *m = strndupa(input_iovec[i].iov_base + STRLEN("MESSAGE="),
-                                           input_iovec[i].iov_len - STRLEN("MESSAGE="));
+                        char *m;
+
+                        m = strndupa_safe(
+                                        (char*) input_iovec[i].iov_base + STRLEN("MESSAGE="),
+                                        input_iovec[i].iov_len - STRLEN("MESSAGE="));
```
Comment (keszybz) [inline src/basic/log.c]: All the other call sites put the first arg on the line with the function call. I think this one should too for consistency.

File: docs/CODING_STYLE.md [hunk 4]
```diff
@@ -365,24 +365,31 @@ SPDX-License-Identifier: LGPL-2.1-or-later
...
 
-- Use `alloca()`, but never forget that it is not OK to invoke `alloca()`
-  within a loop or within function call parameters. `alloca()` memory is
-  released at the end of a function, and not at the end of a `{}` block. Thus,
-  if you invoke it in a loop, you keep increasing the stack pointer without
-  ever releasing memory again. (VLAs have better behavior in this case, so
-  consider using them as an alternative.)  Regarding not using `alloca()`
-  within function parameters, see the BUGS section of the `alloca(3)` man page.
+- Do not use alloca(), strdupa() or strndupa() directly. Use alloca_safe(),
+  strdupa_safe() or strndupa_safe() instead. (The difference is that the latter
```
Comment (keszybz) [inline docs/CODING_STYLE.md]: Backticks are missing here in various places.

File: src/basic/alloc-util.h [hunk 5]
```diff
@@ -22,20 +22,25 @@ typedef void (*free_func_t)(void *p);
 
 #define new0(t, n) ((t*) calloc((n) ?: 1, sizeof(t)))
 
+#define alloca_safe(n)                                                  \
+        ({                                                              \
+                size_t _nn_ = n;                                        \
+                assert(_nn_ <= ALLOCA_MAX);                             \
+                alloca(_nn_ ?: 1);                                      \
```
Comment (medhefgo) [inline src/basic/alloc-util.h]: Shouldn't this be _nn_ > 0? (Not relying on int to bool conversion if it's not strictly a bool)?

Comment (keszybz) [reply src/basic/alloc-util.h]: We'd have to write `_nn_ > 0 ? _nn_ : 1`, which is cumbersome. We skip this rule in some cases where this allows nicely shorter code.

**Comment by poettering** [comment]
Force pushed a new version. Only suggested changes. Upgrading green label.

**Pull request merged**